import {
    ExceptionFilter,
    Catch,
    ArgumentsHost,
    HttpException,
    Logger,
  } from '@nestjs/common';
  import { Request, Response } from 'express';
  
  @Catch()
  export class AllExceptionsFilter implements ExceptionFilter {
    private readonly logger = new Logger(AllExceptionsFilter.name);
  
    catch(exception: unknown, host: ArgumentsHost) {
      const ctx = host.switchToHttp();
      const response = ctx.getResponse<Response>();
      const request = ctx.getRequest<Request>();

      this.logger.error(
        exception
      );
  
      const status =
        exception instanceof HttpException ? exception.getStatus() : 500;
  
      const message =
        exception instanceof HttpException
          ? exception.getResponse()
          : 'An error occurred from our end please try again later or contact us at xxxxxx';
  
      this.logger.error(
        `Status ${status} Error: ${JSON.stringify(message)}`,
      );
  
      response.status(status).json({
        statusCode: status,
        path: request.url,
        error:message,
      });
    }
  }
  